### PR TITLE
Enumerable feature

### PR DESCRIPTION
**Pull request by apricot12**
**In this project, we've rebuilt the following enumerable methods in Ruby:**

- Each

- Each with index

- Select

- All

- Any

- None

- Count

- Map

- Inject

**Additional info:**

- We also made a multiply_els method to test our my_inject method.

- According to the project requirements, we've made it so that my_map accepts both Procs and Blocks with first preference being Procs.

- Yield and call methods were used everywhere applicable.

### CODE REVIEW

**Review by peterrobert — CHANGES_REQUESTED** [review]
### INVALID :stop_sign: 

HELLO, @apricot12 AND @Byabasaija  GREAT WORK SUBMITTING YOUR PROJECT BUT DUE TO THE FOLLOWING REASONS YOUR PROJECT IS MARKED AS INVALID :arrow_down: 

- [x] Your project doesn't have a readme file. Remember this should be added to every project. Kindly add a readme file and make it descriptive.

- [x] Your pull request message should be descriptive. Kindly add points telling us what exactly you have worked on in this project.

No worries :smile:  just complete the above tasks and submit your project for another review.
### if you have any question kindly reach me on slack @peterrobert
### HAPPY CODING :computer:

File: methods/enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    size.times do |index|
+      yield self[index]
+    end
+    self
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_each:

- when a block is given when self is a range returns the range itself after calling the given block once for each element in self:

```ruby

    range = Range.new(5, 50)
    block = proc { |num| num < (5 + 88) / 2 }
    range.my_each(&block) === range.each(&block) => false

```

File: methods/enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_each_with_index
+    return to_enum(:my_each) unless block_given?
+
+    size.times do |index|
+      yield self[index], index
+    end
+    self
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_each_with_index:

- when a block is given when self is a range returns the range itself after calling the given block once for each element in self:

```ruby

    range = Range.new(5, 50)
    block = proc { |num| num < (5 + 88) / 2 }
    range.my_each_with_index(&block) === range.each_with_index(&block) => false

```

File: methods/enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    answer = []
+    my_each do |item|
+      answer.push(item) if yield item
+    end
+    answer
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_select:

- returns an array containing all elements of enum for which the given block returns a true value::range:

```ruby
    
    range = Range.new(5, 50)
    block = proc { |num| num < (5 + 88) / 2 }
    range.my_select(&block) === range.select(&block) => false

```

File: methods/enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_all?
+    return true unless block_given?
+
+    my_each do |item|
+      return false unless yield item
+    end
+    true
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_all?:

- returns false if the block returns false or nil:

```ruby
    
    range = Range.new(5, 50)
    false_block = proc { |num| num > 88 }
    range.my_all?(&false_block) === range.all?(&false_block) => false

```

- when no block or argument is given returns false when one of the collection members are false or nil:

```ruby
    
    false_array = [nil, false, 1, [], 'hi']
    false_array.my_all? === false_array.all? => false

```

- when a class is passed as an argument returns true if all of the collection is a member of such class::Integer:

```ruby
    
    false_array = [1,2,3,4,32,6]
    array.my_all?(Integer) === array.all?(Integer) => false

```

- when a class is passed as an argument returns false if any of the collection is not a member of such class.

- when a Regex is passed as an argument returns true if all of the collection matches the Regex:

```ruby
    
    words = %w[dog door rod blade]
    words.my_all?(/d/) === words.all?(/d/) => false

```

- when a Regex is passed as an argument returns false if any of the collection does not match the Regex.

- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern:

```ruby
    
    array = [3,4 ,43,2,1]
    array.my_all?(3) === array.all?(3) => false

```

- when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern.

File: methods/enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_any?
+    return true unless block_given?
+
+    my_each do |item|
+      return true unless yield item
+    end
+    false
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_any?:

- returns true if the block ever returns a value other than false or nil.

```ruby

     array = [3,4 ,43,2,1]
     true_block = proc { |num| num <= 55 }
     array.my_any?(&true_block)).to eq(array.any?(&true_block)

```

- returns false if the block ever returns a value false or nil:

```ruby

     array = [3,4 ,43,2,1]
     false_block = proc { |num| num > 55 }
     array.my_any?(&false_block) === array.any?(&false_block) => false

```

- you need to support ranges also here.

- when no block or argument is given returns false if at least one of the collection is not true:

```ruby

     false_array = [nil, false, 1, 2, 3]
     false_array.my_any? === false_array.any? => false

```

- when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer.


- when a Regex is passed as an argument returns true if any of the collection matches the Regex.

- when a Regex is passed as an argument returns false if none of the collection matches the Regex.

- when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern:

```ruby

     words = %w[dog door rod blade]
     words.my_any?('cat') === words.any?('cat') => false

```

File: methods/enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_none?
+    return true unless block_given?
+
+    my_each do |item|
+      return false if yield item
+    end
+    true
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_none?:

- support ranges for my_none?.

- when no block or argument is given returns false only if one of the collection members is true:

```ruby

     true_array = [1, 'hi', true]
     true_array.my_none? === false => false

```

- when a class is passed as an argument returns true if none of the collection is a member of such class.

- when a class is passed as an argument returns false if any of the collection is a member of such class.

- when a Regex is passed as an argument returns true if none of the collection matches the Regex.

- when a Regex is passed as an argument returns false if any of the collection matches the Regex.

- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern.

- when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern.

File: methods/enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      my_each do |item|
+        count += 1 if (yield item) == true
+      end
+    else
+      my_each do
+        count += 1
+      end
+    end
+    count
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb:105]: ## my_count

- you should support ranges also.

File: methods/enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    my_each_with_index do |item, i|
+      self[i] = if arg
+                  arg.call(item)
+                else
+                  yield item
+                end
+    end
+    self
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_map:

- returns a new array with the results of running a given block once for every element in enum.

```ruby

      array.my_map { |num| num < 10 } === array.map { |num| num < 10 } => false

```

- does not mutate the original array

File: methods/enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+  def my_inject(initial = 0)
+    my_each do |item|
+      initial = yield(initial, item)
+    end
+    initial
+  end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## my_inject:

- you should support ranges.

File: methods/enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+  def multiply_els
+    my_inject(1) { |initial, item| initial * item }
+  end
+end
```
Comment (elmejdki) [inline methods/enumerables.rb]: ## multiply_els:

- accepts an array as an argument and multiplies all the elements of the array together using #my_inject.

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 
## Great job so far guys, you have a good start, keep up the good work.

I have left for you some comments below please check them.
_You may find that for some requirements I didn't mention examples please take a look at the comments above them you will see the same examples that apply to it, I didn't want to repeat my self that's all_

### Please make all the required changes before you submit for another code review.
## Happy Coding :tophat: 
_If you have any question or you got stuck don't hesitate to reach out to me on slack._

File: methods/enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+      arr.my_each { |item| return false unless item.is_a? arg }
+      true
+    elsif arg.class == Regexp
+      arr.my_each { |item| return false unless item.class =~ arg }
+      true
+    elsif arg.class != Regexp && arg.class != Class
+      arr.my_each { |item| return false unless item == arg }
+      true
+    end
+  end
```
Comment (SpaYco) [inline methods/enumerables.rb]: - [x] when a Regex is passed as an argument it returns true if all of the collection matches the Regex

```Ruby
 words.my_all?(/d/)
```

expected `true`
            got `false`

File: methods/enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+      arr.my_each { |item| return true if item.is_a? arg }
+      false
+    elsif arg.class == Regexp
+      arr.my_each { |item| return true if item.class =~ arg }
+      false
+    elsif arg.class != Regexp && arg.class != Class
+      arr.my_each { |item| return true if item == arg }
+      false
+    end
+  end
```
Comment (SpaYco) [inline methods/enumerables.rb]: - [x] when a Regex is passed as an argument it returns true if any of the collection matches the Regex

```Ruby
  words.my_any?(/d/)
```

expected `true`
            got `false`

File: methods/enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+      my_each do |item|
+        count += 1 if (yield item) == true
+      end
+    else
+      my_each do
+        count += 1
+      end
+    end
+    count
+  end
```
Comment (SpaYco) [inline methods/enumerables.rb:105]: - [x] when a Regex is passed as an argument it returns false if any of the collection matches the Regex

```Ruby
words.my_none?(/d/)
```
 expected `false`
     got `true`

File: methods/enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+    my_each_with_index do |item, i|
+      self[i] = if arg
+                  arg.call(item)
+                else
+                  yield item
+                end
+    end
+    self
+  end
```
Comment (SpaYco) [inline methods/enumerables.rb]: - [x] it returns a new array with the results of running a given block once for every element in enum.

```Ruby
array.my_map { |num| num < 10 }
     
     NoMethodError:
       undefined method `<' for true:TrueClass
```

- [x] it returns a new array with the results of running a given block once for every element in enum.

```Ruby
Failure/Error: block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
     
     NoMethodError:
       undefined method `<' for true:TrueClass
```

- [x] it  returns a new array with the results of running a given block once for every element in enum::range

```Ruby
  range.my_map(&block)

     NoMethodError:
       undefined method `[]=' for 5..50:Range
```

- [x] it accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.

```Ruby
  Failure/Error: my_proc = proc { |num| num > 10 } 

     NoMethodError:
       undefined method `>' for false:FalseClass

```

- [x] it executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

```Ruby
  Failure/Error: my_proc = proc { |num| num > 10 }

     NoMethodError:
       undefined method `>' for false:FalseClass
```

- [x] it does not mutate the original array

```Ruby
  array.my_select { |num| num + 1 }
```

expected: `[2, 7, 0, 4, 1, 7, 2, 1, 1, 6, 2, 0, 4, 6, 1, 6, 4, 7, 3, 8, 8, 6, 1, 5, 0, 1, 4, 7, 8, 6, 0, 1, 7, 6..., 3, 8, 7, 8, 6, 8, 3, 3, 1, 0, 1, 4, 2, 2, 1, 4, 2, 3, 8, 8, 7, 5, 1, 1, 5, 0, 6, 8, 5, 4, 6, 8, 7]`
got: `[3, 8, 1, 5, 2, 8, 3, 2, 2, 7, 3, 1, 5, 7, 2, 7, 5, 8, 4, 9, 9, 7, 2, 6, 1, 2, 5, 8, 9, 7, 1, 2, 8, 7..., 4, 9, 8, 9, 7, 9, 4, 4, 2, 1, 2, 5, 3, 3, 2, 5, 3, 4, 9, 9, 8, 6, 2, 2, 6, 1, 7, 9, 6, 5, 7, 9, 8]`

File: methods/enumerables.rb [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+    index = 0
+
+    each do |item|
+      accum = block.call(accum, item) unless ignore_first && index.zero?
+      index += 1
+    end
+    accum
+  end
+end
```
Comment (SpaYco) [inline methods/enumerables.rb]: - [x] it raises a "LocalJumpError" when no block or argument is given

```Ruby
 array.my_inject
```
expected `LocalJumpError`
 got `#<ArgumentError: you must provide an symbol or a block> with backtrace`

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

Variables used in testing:

```Ruby
  ARRAY_SIZE = 100
  LOWEST_VALUE = 0
  HIGHEST_VALUE = 9

  array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } 
  block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
  words = %w[dog door rod blade] 
  range = Range.new(5, 50) 
  hash = { a: 1, b: 2, c: 3, d: 4, e: 5 } 
  numbers = [1, 2i, 3.14] 
  array_clone = array.clone 
```

## Required Changes ♻️ 

- [x] Do not make any changes in config files - they represent style guidelines that you share with your team - which is a group of all Microverse students. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/linter_config.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/54883773/))

- [x] more below 👇 



### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

File: .rubocop.yml [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+Layout/LineLength:
+  Max: 120
+Metrics/MethodLength:
+  Max: 100
+Metrics/ModuleLength:
+  Max: 200
+Metrics/PerceivedComplexity:
+  Enabled: false
+Metrics/CyclomaticComplexity:
+  Enabled: false
```
Comment (kitpao) [inline .rubocop.yml]: - [ ] Please, do not make changes to the rubocop file. Here you have the [original rubocop from Microverse](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml). Kindly use the original with no modifications
image

File: .rubocop.yml [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+Style/ClassAndModuleChildren:
+  Enabled: false
+Style/EachForSimpleLoop:
+  Enabled: false
+Style/AndOr:
+  Enabled: false
+Style/DefWithParentheses:
+  Enabled: false
+Style/For:
+  Enabled: false
```
Comment (kitpao) [inline .rubocop.yml]: - [ ] Please, do not make changes to the rubocop file. Here you have the [original rubocop from Microverse](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml). Kindly use the original with no modifications
image

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @Byabasaija and @apricot12 

Great job fixing all your methods! 🥳 All of them are passing

But remember that in Microverse (and any other established company you might work in) you will have to follow guidelines, in this case, the linter files. Please do not modify them. I leave you the link to the original below 🔽 

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Comment by kitpao** [comment]
Note: Please be so kind as not to make changes after requesting a code review :( I already submitted it to microverse

**Review by kitpao — APPROVED** [review]
# 👋 Hello again @Byabasaija and @apricot12 

### 👏 Your project has been approved 👏
### 🚀 You may merge to develop and submit the project completition form 🚀

___
##### 👉 If you have any doubts you can contact me on slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate it if you could leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😄  

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Pull request merged**